Update dotnet SDK to 6.0 Preview 4

ViktorHofer opened the pull request:
Update the minimum and target version of the SDK to 6.0 Preview 4.

Part of https://github.com/dotnet/runtime/issues/53411

ViktorHofer committed:
Update dotnet SDK to 6.0 Preview 4

Update the minimum and target version of the SDK to 6.0 Preview 4.

Part of https://github.com/dotnet/runtime/issues/53411

ViktorHofer committed:
Update scripts.csproj

ViktorHofer committed:
Update scripts.csproj

akoeplinger committed:
Switch to using eng/targetingpacks.targets for redirecting runtime pack location in src/mono/sample

ViktorHofer commented:
A loader test is now failing consistently with the SDK upgrade:

```
Loader/binding/tracing/BinderTracingTest.Basic/BinderTracingTest.Basic.sh

Return code:      1
Raw output file:      /tmp/helix/working/AEF7094F/w/A25509A3/uploads/Reports/Loader.binding/tracing/BinderTracingTest.Basic/BinderTracingTest.Basic.output.txt
Raw output:
BEGIN EXECUTION
/tmp/helix/working/AEF7094F/p/corerun BinderTracingTest.Basic.dll ''
[4:47:35 AM] Running LoadFile...
[4:47:35 AM] Running LoadBytes...
[4:47:36 AM] Running LoadFromStream...
[4:47:36 AM] Running LoadFromAssemblyPath...
[4:47:36 AM] Launching process for LoadFromAssemblyName...
[4:47:36 AM] Running LoadFromAssemblyName...

[4:47:37 AM] Launching process for LoadFrom...
[4:47:37 AM] Running LoadFrom...

[4:47:37 AM] Launching process for PlatformAssembly...
[4:47:37 AM] Running PlatformAssembly...

[4:47:38 AM] Launching process for PlatformAssembly_Cached...
[4:47:38 AM] Running PlatformAssembly_Cached...

[4:47:38 AM] Launching process for Reflection...
[4:47:38 AM] Running Reflection...

[4:47:39 AM] Launching process for Reflection_Cached...
[4:47:39 AM] Running Reflection_Cached...

[4:47:40 AM] Launching process for Reflection_CustomALC...
[4:47:40 AM] Running Reflection_CustomALC...

[4:47:40 AM] Launching process for ContextualReflection_DefaultToCustomALC...
[4:47:40 AM] Running ContextualReflection_DefaultToCustomALC...

[4:47:41 AM] Launching process for ContextualReflection_CustomToDefaultALC...
[4:47:41 AM] Running ContextualReflection_CustomToDefaultALC...

[4:47:41 AM] Launching process for JITLoad...
[4:47:41 AM] Running JITLoad...

[4:47:42 AM] Launching process for JITLoad_CustomALC...
[4:47:42 AM] Running JITLoad_CustomALC...

[4:47:42 AM] Running AssemblyInAppPath...
[4:47:43 AM] Running NonExistentAssembly...
[4:47:43 AM] Launching process for SatelliteAssembly_AppPath...
[4:47:43 AM] Running SatelliteAssembly_AppPath...

[4:47:43 AM] Running SatelliteAssembly_CultureSubdirectory...
Test SatelliteAssembly_CultureSubdirectory failed: System.IO.FileNotFoundException: Could not load file or assembly 'AssemblyToLoad_Subdirectory.resources, Culture=fr-FR, PublicKeyToken=null'. The system cannot find the file specified.

File name: 'AssemblyToLoad_Subdirectory.resources, Culture=fr-FR, PublicKeyToken=null'
at System.Reflection.RuntimeAssembly.InternalLoad(ObjectHandleOnStack assemblyName, ObjectHandleOnStack requestingAssembly, StackCrawlMarkHandle stackMark, Boolean throwOnFileNotFound, ObjectHandleOnStack assemblyLoadContext, ObjectHandleOnStack retAssembly) in System.Private.CoreLib.dll:token 0x6004cd3+0x8
at System.Reflection.RuntimeAssembly.InternalLoad(AssemblyName assemblyName, RuntimeAssembly requestingAssembly, StackCrawlMark& stackMark, Boolean throwOnFileNotFound, AssemblyLoadContext assemblyLoadContext) in System.Private.CoreLib.dll:token 0x6004cd2+0x16
at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyName(AssemblyName assemblyName) in System.Private.CoreLib.dll:token 0x6003564+0xe
at BinderTracingTests.BinderTracingTest.SatelliteAssembly_CultureSubdirectory() in BinderTracingTest.Basic.dll:token 0x6000013+0x33
at BinderTracingTests.BinderTracingTest.RunSingleTest(MethodInfo method) in BinderTracingTest.Basic.dll:token 0x600001c+0xd8
Expected: 100
Actual: 101
END EXECUTION - FAILED
Test Harness Exitcode is : 1
To run the test:

set CORE_ROOT=/tmp/helix/working/AEF7094F/p
/private/tmp/helix/working/AEF7094F/w/A25509A3/e/Loader/binding/tracing/BinderTracingTest.Basic/BinderTracingTest.Basic.sh
Expected: True
Actual:   False
```

@jkotas @BruceForstall @agocke who would be the best person to take a look?

BruceForstall commented:
Looks like @elinor-fung knows this test

elinor-fung commented:
The binder test uses project references and `SatelliteDllsProjectOutputGroup` to copy the satellite DLL from another project to its output directory. It then copies that satellite DLL from its output directory into a specific subdirectory.

With Preview 4, it looks like the satellite DLL is now placed in a subdirectory of the test project's output directory (instead of directly in the output directory) as a result of a change in the `AssignTargetsPath` task: https://github.com/dotnet/msbuild/pull/6237

Will look at making the test work with this.

elinor-fung committed:
Fix binder tracing tests

akoeplinger committed:
Set UseMonoRuntime property so the dotnet SDK can pull the mono runtime packs

This is necessary after the changes from https://github.com/dotnet/sdk/pull/16909.

ViktorHofer commented on eng/Subsets.props:
```diff
@@ -39,6 +39,9 @@
     <RuntimeFlavor Condition="'$(RuntimeFlavor)' == ''">CoreCLR</RuntimeFlavor>
+
+    <!-- this property is used by the SDK to pull in mono-based runtime packs -->
+    <UseMonoRuntime Condition="'$(UseMonoRuntime)' == '' and '$(RuntimeFlavor)' == 'Mono'">true</UseMonoRuntime>
```
can you please move that property out of Subsets.props into Directory.Build.props to the feature switches? https://github.com/dotnet/runtime/blob/b52d276e3c18687a120747e2edfd3ae50679683f/Directory.Build.props#L205

akoeplinger replied:
Done. I couldn't put it in Feature switches since it needs the Subsets.props import which is after that one.

akoeplinger committed:
Move UseMonoRuntimeProperty and set it to false in coreclr

lambdageek commented on src/mono/sample/HelloWorld/HelloWorld.csproj:
```diff
@@ -2,16 +2,5 @@
-    <EnableTargetingPackDownload>false</EnableTargetingPackDownload>
   </PropertyGroup>
-
-  <!-- Redirect 'dotnet publish' to in-tree runtime pack -->
-  <Target Name="TrickRuntimePackLocation" AfterTargets="ProcessFrameworkReferences">
-    <ItemGroup>
-      <RuntimePack>
-        <PackageDirectory>$(ArtifactsBinDir)microsoft.netcore.app.runtime.$(RuntimeIdentifier)\$(Configuration)</PackageDirectory>
-      </RuntimePack>
-    </ItemGroup>
-    <Message Text="Packaged ID: %(RuntimePack.PackageDirectory)" Importance="high" />
-  </Target>
```
@ViktorHofer 

This broke the mono samples.   Am I missing something?  (I did: `./build.sh mono+libs -c Release ` to build)

```
$ make run MONO_CONFIG=Release
../../../.././dotnet.sh publish -c Release -r osx-x64
Microsoft (R) Build Engine version 17.0.0-preview-21302-02+018bed83d for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
  Used runtime pack:
  Used runtime pack: /Users/alklig/.nuget/packages/microsoft.aspnetcore.app.runtime.osx-x64/6.0.0-preview.5.21301.17
/usr/local/share/dotnet/sdk/6.0.100-preview.5.21302.13/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(391,5): error NETSDK1112: The runtime pack for Microsoft.NETCore.App.Runtime.Mono.osx-x64 was not downloaded. Try running a NuGet restore with the RuntimeIdentifier 'osx-x64'. [/Users/alklig/work/dotnet-runtime/delta-bugs/src/mono/sample/HelloWorld/HelloWorld.csproj]
make: *** [publish] Error 1
```
